library/core/lifecycle-events/src/main/java/org/quiltmc/qsl/lifecycle/api/event/ServerLifecycleEvents.java

```diff
@@ -0,0 +1,170 @@
+	 * and players are still connected to the server.
+	 *
+	 * <h2>What should mods do when this event is executed</h2>
+	 *
+	 * Mods may do clean up work when this event is executed, such as shutting down any asynchronous executors,
+	 * databases and saving auxiliary mod data.
+	 */
+	public static final ArrayEvent<Stopping> STOPPING = ArrayEvent.create(Stopping.class, callbacks -> server -> {
+		for (var callback : callbacks) {
+			callback.stoppingServer(server);
+		}
+	});
+
+	/**
+	 * An event indicating that a Minecraft server has finished shutting down.
+	 *
+	 * <p>What occurs after this event will vary depending on whether the Minecraft server is a dedicated server or the
+	 * integrated server of a Minecraft client:
+	 *
+	 * <ul>
+	 * <li><b>integrated server:</b> the client will continue to tick after this event is executed. If the client is being
+	 * shut down, then this event is called after events indicating the client is being shut down are called.
+	 * <li><b>dedicated server:</b> this will be the last event called before the Java Virtual Machine terminates.
+	 * </ul>
+	 *
+	 * <h2>What should mods do when this event is executed?</h2>
+	 *
+	 * Mods should stop referencing this Minecraft server or else the dead server will continue to be tracked on the
+	 * heap and will leak memory. Though this doesn't matter when the server is a dedicated server, it is good principle
+	 * to clean up references you no longer need regardless.
+	 */
+	public static final ArrayEvent<Stopped> STOPPED = ArrayEvent.create(Stopped.class, callbacks -> server -> {
+		for (var callback : callbacks) {
+			callback.exitServer(server);
+		}
+	});
+
+	private ServerLifecycleEvents() {}
+
+	/**
+	 * Functional interface to be implemented on callbacks for {@link #STARTING}.
+	 * @see #STARTING
+	 */
+	@FunctionalInterface
+	public interface Starting {
+		/**
+		 * Called when a Minecraft server is starting.
+		 *
+		 * @param server the server which is starting
+		 */
+		void startingServer(MinecraftServer server);
+	}
+
+	/**
+	 * Functional interface to be implemented on callbacks for {@link #READY}.
+	 * @see #READY
+	 */
+	@FunctionalInterface
+	public interface Ready {
+		/**
+		 * Called when a Minecraft server is ready to tick and accept players.
+		 *
+		 * @param server the server which is ready
+		 */
+		void readyServer(MinecraftServer server);
+	}
+
+	/**
+	 * Functional interface to be implemented on callbacks for {@link #STOPPING}.
+	 * @see #STOPPING
+	 */
+	@FunctionalInterface
+	public interface Stopping {
+		/**
+		 * Called when a Minecraft server has finished it's last tick and is shutting down.
+		 *
+		 * @param server the server which is shutting down
+		 */
+		void stoppingServer(MinecraftServer server);
+	}
+
+	/**
+	 * Functional interface to be implemented on callbacks for {@link #STOPPED}.
+	 * @see #STOPPED
+	 */
+	@Environment(EnvType.CLIENT)
```

Comment:
That... doesn't seem correct, considering the event field isn't also marked as client-only.

Reply:
I thought I got these all 😢